builtin/providers/openstack/resource_openstack_blockstorage_volume_v1.go

```diff
@@ -0,0 +1,347 @@
+			},
+			"attachment": &schema.Schema{
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"instance_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"device": &schema.Schema{
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+				Set: resourceVolumeAttachmentHash,
+			},
+		},
+	}
+}
+
+func resourceBlockStorageVolumeV1Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	blockStorageClient, err := config.blockStorageV1Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+
+	createOpts := &volumes.CreateOpts{
+		Description: d.Get("description").(string),
+		Name:        d.Get("name").(string),
+		Size:        d.Get("size").(int),
+		SnapshotID:  d.Get("snapshot_id").(string),
+		SourceVolID: d.Get("source_vol_id").(string),
+		ImageID:     d.Get("image_id").(string),
+		VolumeType:  d.Get("volume_type").(string),
+		Metadata:    resourceContainerMetadataV2(d),
+	}
+
+	log.Printf("[INFO] Requesting volume creation")
+	v, err := volumes.Create(blockStorageClient, createOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack volume: %s", err)
+	}
+	log.Printf("[INFO] Volume ID: %s", v.ID)
+
+	// Store the ID now
+	d.SetId(v.ID)
+
+	// Wait for the volume to become available.
+	log.Printf(
+		"[DEBUG] Waiting for volume (%s) to become available",
+		v.ID)
+
+	stateConf := &resource.StateChangeConf{
+		Target:     "available",
+		Refresh:    VolumeV1StateRefreshFunc(blockStorageClient, v.ID),
+		Timeout:    10 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for volume (%s) to become ready: %s",
+			v.ID, err)
+	}
+
+	return resourceBlockStorageVolumeV1Read(d, meta)
+}
+
+func resourceBlockStorageVolumeV1Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	blockStorageClient, err := config.blockStorageV1Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+
+	v, err := volumes.Get(blockStorageClient, d.Id()).Extract()
+	if err != nil {
+		return CheckDeleted(d, err, "volume")
+	}
+
+	log.Printf("\n\ngot volume: %+v\n\n", v)
+
+	log.Printf("[DEBUG] Retreived volume %s: %+v", d.Id(), v)
+
+	d.Set("region", d.Get("region").(string))
+	d.Set("size", v.Size)
+
+	if t, exists := d.GetOk("description"); exists && t != "" {
+		d.Set("description", v.Description)
+	} else {
+		d.Set("description", "")
+	}
+
+	if t, exists := d.GetOk("name"); exists && t != "" {
+		d.Set("name", v.Name)
+	} else {
+		d.Set("name", "")
+	}
```

Comment:
It looks like these patterns are intending to "only set the value once", yeah?

In other providers we've tended to just do `d.Set()` without any guarding. This has the benefit of allowing terraform to "catch up" with some out-of-band changes to the resource, and also simplifying the code a little.

Let me know if there's an important reason for these guards I'm missing, though. :+1: